### PR TITLE
Basics of object allocation reporting - just user objects for now

### PR DESCRIPTION
**Pull request by chrisseaton**
We want to report all allocations - but we're using the layout DSL which hides it from us!

We could:

* Manually report allocations around every use of the layout DSL allocate method
* Write a facade that reports the allocations then calls the layout DSL
* Modify the layout DSL to report allocations

We probably need to work out what is wanted from the users of this API more, and to think more about size estimation, before we do any work.

### CODE REVIEW

**Comment by eregon** [comment]
We could use the [`build` Alternative constructor method](https://graalvm.github.io/graal/truffle/javadoc/com/oracle/truffle/api/object/dsl/Layout.html) of the Layout DSL exclusively, which makes it then much simpler to use an AllocateObjectNode or so around to track allocations.
We could use a AllocateObjectNode even when we know the factory, or maybe just have a wrapping method taking the factory, arguments and also track allocations.

**Comment by chrisseaton** [comment]
Ah, `build` is probably what we want. Did I write that? I can't remember.

We don't want to make people use a node as it's often not easy to do so - hence the static DSL in the first place.

File: src/main/java/org/truffleruby/language/objects/AllocateObjectNode.java [hunk 1]
```diff
@@ -170,4 +171,20 @@ protected int getCacheLimit() {
         return getContext().getOptions().ALLOCATE_CLASS_CACHE;
     }
 
+    private DynamicObject allocate(DynamicObjectFactory factory, Object[] values) {
+        final AllocationReporter allocationReporter = getContext().getAllocationReporter();
+
+        if (allocationReporter.isActive()) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/objects/AllocateObjectNode.java]: Is this check guaranteed to fold?

Comment (chrisseaton) [inline src/main/java/org/truffleruby/language/objects/AllocateObjectNode.java]: Yes.

**Comment by eregon** [comment]
I wrote `build` for more felxibility :)
We don't necessarily need a node but then the allocationReporter.isActive() check will cost a bit of course.

**Pull request merged**